modules/dnsmasq_dhcp/autodetection.go

```diff
@@ -128,3 +217,85 @@ func parseDHCPRangeLine(s string) (r string) {
 
 	return fmt.Sprintf("%s-%s", start, end)
 }
+
+func (dir configDir) findConfigs() ([]string, error) {
+	fis, err := ioutil.ReadDir(dir.path)
+	if err != nil {
+		return nil, err
+	}
+
+	var configs []string
+
+	for _, fi := range fis {
+		if !fi.Mode().IsRegular() {
+			continue
+		}
+
+		// We copy the dnsmasq's logic
+		//
+		// /* ignore emacs backups and dotfiles */
+		// if (len == 0 ||
+		//     ent->d_name[len - 1] == '~' ||
+		//     (ent->d_name[0] == '#' && ent->d_name[len - 1] == '#') ||
+		//     ent->d_name[0] == '.')
+		//    continue;
+
+		name := fi.Name()
+		if strings.HasSuffix(name, "~") ||
+			(strings.HasPrefix(name, "#") && strings.HasSuffix(name, "#")) ||
+			strings.HasPrefix(name, ".") {
+			continue
+		}
+
+		if len(dir.includeSuffix) > 0 {
+			including := false
+			for _, suffix := range dir.includeSuffix {
+				if strings.HasSuffix(fi.Name(), suffix) {
+					including = true
+					break
+				}
+			}
+
+			if !including {
+				continue
+			}
+		}
+
+		excluding := false
+		for _, suffix := range dir.excludeSuffix {
+			if strings.HasSuffix(fi.Name(), suffix) {
+				excluding = true
+				break
+			}
+		}
```

Comment:
this is getting complicated, i suggest to simplify, it looks like it is time to wrap []string (includeSuffix, excludeSuffix) to its own type and add methods to it, after it it will be somewthing like that

```go
for _, fi := range fis {
    if !fi.Mode().IsRegular() || isBadFileName(fi.Name()){
        continue
    }

    if !dir.inlcudeSuffic.match(fi.Name()) || dir.excludeSuffic.match(fi.Name()) {
        continue
    }
    ...
    ...  
```

nested if/else with break is hard to read/follow, please simplify

Reply:
^^ it is likely i am wrong with logic, please check it, code snippet above is just to show the idea

Reply:
i think it should be

```go
if dir.matchConfig(fi.Name()) { ... }
```

because it should match every file if both includeSuffinx and excludeSuffix is nil

Reply:
I think two dead simple loops have much less cognitive load than having two separated types which will be used in one place. Moreover, adding separated types will make logic fragmented.

Reply:
```go
for _, fi := range fis {
    if !fi.Mode().IsRegular() {
        continue
    }

    isBadFileName(fi.Name()) {
         continue
     }

     if !dir.matchConfig(fi.Name())  {
         continue
     }
```

vs

```go
	for _, fi := range fis {
		if !fi.Mode().IsRegular() {
			continue
		}

		// We copy the dnsmasq's logic
		//
		// /* ignore emacs backups and dotfiles */
		// if (len == 0 ||
		//     ent->d_name[len - 1] == '~' ||
		//     (ent->d_name[0] == '#' && ent->d_name[len - 1] == '#') ||
		//     ent->d_name[0] == '.')
		//    continue;

		name := fi.Name()
		if strings.HasSuffix(name, "~") ||
			(strings.HasPrefix(name, "#") && strings.HasSuffix(name, "#")) ||
			strings.HasPrefix(name, ".") {
			continue
		}

		if len(dir.includeSuffix) > 0 {
			including := false
			for _, suffix := range dir.includeSuffix {
				if strings.HasSuffix(fi.Name(), suffix) {
					including = true
					break
				}
			}

			if !including {
				continue
			}
		}

		excluding := false
		for _, suffix := range dir.excludeSuffix {
			if strings.HasSuffix(fi.Name(), suffix) {
				excluding = true
				break
			}
		}

		if excluding {
			continue
		}

		configs = append(configs, filepath.Join(dir.path, fi.Name()))
	}

```

Reply:
in 1 case we have only logic, in 2nd logic + details. You can add `matchConfig` to `configDir` struct. Whatever you want.

Reply:
https://github.com/netdata/go.d.plugin/pull/239#discussion_r301149374

Reply:
yes, i saw, nice, thanks :+1:

Reply:
Oh, please dont force push on every commit, we will squash them during merge

Reply:
I pushed since you insist, but I don't like that changeset because it makes harder to ensure that the module follows dnsmasq's logic correctly.

Reply:
Why is it harder now? What is changed? We moved implementation details to its own method which can be tested afther those changes.

Reply:
So your point that before it was copy of `c` code and now it is not?

Reply:
I found the code from the first iteration much more convenient for side-by-side checking logic with dnsmasq's code.

Reply:
Let me know if you have any questions/issues with the code.

Reply:
Ok, i got your point.